### ISSUE TITLE
Documentation of possible options for **task**

### ISSUE THREAD

**Issue by jotix16** [rwth-i6/returnn#471]
Under [\_\_main\_\_.py ](https://github.com/rwth-i6/returnn/blob/master/returnn/__main__.py) in [execute_main_task()](https://github.com/rwth-i6/returnn/blob/master/returnn/__main__.py#LC440:~:text=execute_main_task,-()%3A) RETURNN decides what task is performed.  The source code in [execute_main_task()]() is not documented and some tasks are only briefly mentioned in [execution guide](https://returnn.readthedocs.io/en/latest/getting_started/tech_overview.html#execution-guide). 

While several tasks such as train and eval are run most often it would be nice to document and test out the rest of them too.
I believe there is a lot of potential to be unlocked if all possible tasks are clearly documented. Maybe several tasks don't even run as expected and could be cleaned out.

For example, for a novice like me it is complicated to get a pipeline up and running with the first try. 

If a certain task could automatically test the network with a dummy input or test/analyse the dataset it would make the life much easier. One could try them out separately and put them together afterwards. Maybe this would be helpful to proof check some of the external imports discussed in #436, too. 
I am nor certain neither clear how this could be done or if it is possible at all.

The list of tasks as of right now is:
### used
- [x]   **train**: Trains the network with the given dataset. It requires at least train = train_dataset_dict. If eval, dev or eval_datasets are specified they are evaluated at the end of each epoch.
- [x]   **eval**:  Evaluates eval, dev or eval_datasets if specified
- [ ]   **forward, hpx**:
- [ ]   **search**: 
- [x]   **nop**: Beside dataset and network, initializes the rest
- [x]   **nop_init_net_train**: Initializes model and training data but doesnt start training
- [x]   **initialize_model**: Similiar to nop_init_net_train but it saves a checkpoint at the end
- [x]   **cleanup_old_models**: Clean up models if we have done some lr control. with cleanup_old_models="" we can specify more options.
- [x]   **compute_priors**: Computes the priors of network outputs(or posterior?) for the training dataset.
- [x]   **analyze**: Analyses training dataset for the given network. Calculates stuff such as loss, perplexity, ce, frame error, seq length, and prob histograms per batch and for one whole epoch(accumulated).
- [ ]   **hyper_param_tuning**: 
- [ ]   **search_server**: Starts a web-server with a simple API to forward data through the network (not sure what/why/how)
- [x]   **config:"action_name"**:  runs the action_name(), action musst be callable
- [x]   **optional-config:"action_name"**:  similiar to config but the action can be None too

### not working
- [ ] **analyze_data**: Doesn't initialize the network at all. For the dataset given by analyze_dataset="train" it calculates mean, priors and std and dumps them in seperate files with prefix given by stat_prefix="path/to/prefix".
  Problem with ```inplace_increment```, requires theano or a global ```_have_inplace_increment??```. ```assert not ds.is_data_sparse(data_key), "needed for mean/var estimation"``` not should go away?

### deprecated(theano)
- [ ]   **theano_graph**:
- [ ]   **classify**: 'Engine' object has no attribute 'classify' (theano only)
- [ ]   **server**:
- [ ]   **daemon**:

**Comment by albertz** [comment]
Note that we also **want** that people look at the code and not just the documentation. RETURNN is not intended for users who do not want to look at its code.

Also note that many of these tasks are somewhat outdated (and maybe broken) and should rather be removed instead of being documented. (I'm always hesitating to document totally outdated or deprecated features...)
So better only document what you think is meaningful (and looks not broken).

**Comment by jotix16** [comment]
> Note that we also want that people look at the code and not just the documentation. RETURNN is not intended for users who do not want to look at its code.

Are you saying, that if things are well documented, people will not look at the code? If **contribution** is what you intend when you say *looking at the code*, I think documentation makes it only easier. 

For example, I want to understand/improve ```A()``` which calls ```B()```. ```B()``` is not important for the logic of the function but is ambiguous what it exactly does(especially for global variables, such as config). Then, I have to go through the source code of ```B()``` too. If ```B()``` would have been documented instead, the process would have been more efficient.

> tasks are somewhat outdated (and maybe broken) and should rather be removed instead of being documented

Exactly, I noticed that.

**Comment by jotix16** [comment]
> This is not a strict rule, of course. There will always be some overlap. But this is an example where I'm not sure why the documentation would be helpful because you can trivially see that already from the code.

You are right. It is a little hard to see what requirements each task has but don't think it has a solution beside checking the code. 

> Function documentation should be in the docstrings. So when you browse the code in an editor, you will directly see the docstring documentation of functions (and classes, modules).

I meant that actually.

**Comment by albertz** [comment]
I see that you checked some of the tasks. Did you work on some documentation? There can be some documentation if you think it is helpful, but I think maybe more as an overview, and only the important tasks. If you prepared some documentation, maybe submit it as a PR.

**Comment by jotix16** [comment]
I didn't look into documentation further. The stuff I wrote is on the checklist. They are more like motivation for each task.
Should I add them as documentation under `execute_main_task()` ?

I think that some minimal requirements for each task are also necessary as a startpoint. Or is it too much?

**Comment by albertz** [comment]
For people looking already at this function `execute_main_task`, I think this is not needed. Unless some task is maybe unclear, then maybe you can add a comment there in the code.

I was thinking about somewhere in the generic documentation. But only if you see this useful or necessary.

> I think that some minimal requirements for each task are also necessary as a startpoint.

What do you mean exactly? Like that you need a network + dataset for the `train` task?

**Comment by jotix16** [comment]
I think it is useful, added it under [execution-guide](https://returnn.readthedocs.io/en/latest/getting_started/tech_overview.html#execution-guide) in #480.

> What do you mean exactly? Like that you need a network + dataset for the train task?

Yes. For example, I am able to run ``train`` and ``eval``,but not ``search`` on the [demo-tf-enc-dec.config ](https://github.com/rwth-i6/returnn/blob/master/demos/demo-tf-enc-dec.config).

**Comment by albertz** [comment]
> I think it is useful, added it under [execution-guide](https://returnn.readthedocs.io/en/latest/getting_started/tech_overview.html#execution-guide) in #480.

Thanks, merged.

> For example, I am able to run `train` and `eval`,but not `search` on the [demo-tf-enc-dec.config ](https://github.com/rwth-i6/returnn/blob/master/demos/demo-tf-enc-dec.config).

This sounds like a bug in the config? Or because `search_data` is missing? Idk. I think it should work. Maybe you can open a separate issue about that.
